code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/internal/EdgePersonalizationResponseHandler.java

```diff
@@ -163,128 +191,207 @@ void fetchMessages(final List<Surface> surfaces) {
+            }
+
+            for (final PropositionItem propositionItem : proposition.getItems()) {
+                JSONObject ruleJson;
+                try {
+                    ruleJson = new JSONObject(propositionItem.getContent());
+                } catch (final JSONException jsonException) {
+                    Log.debug(MessagingConstants.LOG_TAG, SELF_TAG, "processPropositions - Skipping proposition with invalid json content.");
                     continue;
                 }
 
-                for (final PayloadItem payloadItem : proposition.items) {
-                    final JSONObject ruleJson = payloadItem.data.getRuleJsonObject();
-                    if (ruleJson == null) {
-                        Log.debug(MessagingConstants.LOG_TAG, SELF_TAG, "processPropositions - Skipping proposition with no in-app message content.");
-                        continue;
-                    }
+                final List<LaunchRule> parsedRule = JSONRulesParser.parse(propositionItem.getContent(), extensionApi);
+                if (parsedRule == null) {
+                    Log.debug(MessagingConstants.LOG_TAG, SELF_TAG, "Skipping proposition with malformed rule json content.");
+                    continue;
+                }
 
-                    final List<LaunchRule> parsedRule = JSONRulesParser.parse(ruleJson.toString(), extensionApi);
-                    if (parsedRule == null) {
-                        Log.debug(MessagingConstants.LOG_TAG, SELF_TAG, "Skipping proposition with malformed in-app message content.");
+                final RuleConsequence consequence = parsedRule.get(0).getConsequenceList().get(0);
+                // store reporting data for this payload for later use
+                final String messageId = consequence.getId();
+                if (!StringUtils.isNullOrEmpty(messageId)) {
+                    final PropositionInfo propositionInfo = PropositionInfo.createFromProposition(proposition);
+                    if (propositionInfo == null) {
+                        Log.debug(MessagingConstants.LOG_TAG, SELF_TAG, "Skipping proposition with missing / invalid proposition info.");
                         continue;
                     }
+                    tempPropositionInfo.put(messageId, PropositionInfo.createFromProposition(proposition));
+                }
 
+                final boolean isInAppConsequence = MessagingUtils.isInApp(consequence);
+                List<Proposition> propositionsForSurface;
+                if (isInAppConsequence) {
+                    propositionsForSurface = inAppPropositions.get(surface) == null ? new ArrayList<>() : inAppPropositions.get(surface);
+                    if (!propositionsForSurface.isEmpty()) {
+                        inAppPropositions.get(surface).add(proposition);
+                    } else {
+                        propositionsForSurface.add(proposition);
+                        inAppPropositions.put(surface, propositionsForSurface);
```

Comment:
nit: We should have a util for this (similar to iOS) as we are doing it for inapp, feed items etc